File: backend/FantasyLeague.Api/Models/ScoringSettings.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace FantasyLeague.Api.Models
{
    public class ScoringSettings
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public int LeagueId { get; set; }

        [ForeignKey("LeagueId")]
        public virtual League? League { get; set; }

        [Required]
        [StringLength(10)]
        public string Sport { get; set; } = "NFL"; // NFL, NBA, MLB

        // Quarterback Scoring
        [Column(TypeName = "decimal(5,2)")]
        public decimal PassingYardsPerPoint { get; set; } = 25.0m; // 1 point per 25 yards

        [Column(TypeName = "decimal(5,2)")]
        public decimal PassingTouchdownPoints { get; set; } = 4.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal PassingInterceptionPoints { get; set; } = -1.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal PassingTwoPointConversion { get; set; } = 2.0m;

        // Rushing Scoring
        [Column(TypeName = "decimal(5,2)")]
        public decimal RushingYardsPerPoint { get; set; } = 10.0m; // 1 point per 10 yards

        [Column(TypeName = "decimal(5,2)")]
        public decimal RushingTouchdownPoints { get; set; } = 6.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal RushingTwoPointConversion { get; set; } = 2.0m;

        // Receiving Scoring
        [Column(TypeName = "decimal(5,2)")]
        public decimal ReceivingYardsPerPoint { get; set; } = 10.0m; // 1 point per 10 yards

        [Column(TypeName = "decimal(5,2)")]
        public decimal ReceivingTouchdownPoints { get; set; } = 6.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal ReceptionPoints { get; set; } = 0.5m; // Half-PPR default

        [Column(TypeName = "decimal(5,2)")]
        public decimal ReceivingTwoPointConversion { get; set; } = 2.0m;

        // General Offensive
        [Column(TypeName = "decimal(5,2)")]
        public decimal FumbleLostPoints { get; set; } = -2.0m;

        // Kicker Scoring
        [Column(TypeName = "decimal(5,2)")]
        public decimal ExtraPointPoints { get; set; } = 1.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal FieldGoal0to39Points { get; set; } = 3.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal FieldGoal40to49Points { get; set; } = 4.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal FieldGoal50PlusPoints { get; set; } = 5.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal MissedExtraPointPoints { get; set; } = -1.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal MissedFieldGoalPoints { get; set; } = -1.0m;

        // Defense/Special Teams Scoring
        [Column(TypeName = "decimal(5,2)")]
        public decimal DefenseTouchdownPoints { get; set; } = 6.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal SackPoints { get; set; } = 1.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal InterceptionPoints { get; set; } = 2.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal FumbleRecoveryPoints { get; set; } = 2.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal SafetyPoints { get; set; } = 2.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal BlockedKickPoints { get; set; } = 2.0m;

        // Defense Points Allowed (Yahoo standard)
        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed0Points { get; set; } = 10.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed1to6Points { get; set; } = 7.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed7to13Points { get; set; } = 4.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed14to20Points { get; set; } = 1.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed21to27Points { get; set; } = 0.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed28to34Points { get; set; } = -1.0m;

        [Column(TypeName = "decimal(5,2)")]
        public decimal DefensePointsAllowed35PlusPoints { get; set; } = -4.0m;

        // Additional settings
        [Column(TypeName = "decimal(5,2)")]
        public decimal BenchPoints { get; set; } = 0.0m; // Points for bench players

        public bool IsActive { get; set; } = true;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

        // Helper method to get Yahoo default settings
        public static ScoringSettings GetYahooDefaults(int leagueId)
        {
            return new ScoringSettings
            {
                LeagueId = leagueId,
                Sport = "NFL",
                // Passing
                PassingYardsPerPoint = 25.0m,
                PassingTouchdownPoints = 4.0m,
                PassingInterceptionPoints = -1.0m,
                PassingTwoPointConversion = 2.0m,
                // Rushing
                RushingYardsPerPoint = 10.0m,
                RushingTouchdownPoints = 6.0m,
                RushingTwoPointConversion = 2.0m,
                // Receiving (Half-PPR)
                ReceivingYardsPerPoint = 10.0m,
                ReceivingTouchdownPoints = 6.0m,
                ReceptionPoints = 0.5m,
                ReceivingTwoPointConversion = 2.0m,
                // General
                FumbleLostPoints = -2.0m,
                // Kicking
                ExtraPointPoints = 1.0m,
                FieldGoal0to39Points = 3.0m,
                FieldGoal40to49Points = 4.0m,
                FieldGoal50PlusPoints = 5.0m,
                MissedExtraPointPoints = -1.0m,
                MissedFieldGoalPoints = -1.0m,
                // Defense
                DefenseTouchdownPoints = 6.0m,
                SackPoints = 1.0m,
                InterceptionPoints = 2.0m,
                FumbleRecoveryPoints = 2.0m,
                SafetyPoints = 2.0m,
                BlockedKickPoints = 2.0m,
                // Defense Points Allowed
                DefensePointsAllowed0Points = 10.0m,
                DefensePointsAllowed1to6Points = 7.0m,
                DefensePointsAllowed7to13Points = 4.0m,
                DefensePointsAllowed14to20Points = 1.0m,
                DefensePointsAllowed21to27Points = 0.0m,
                DefensePointsAllowed28to34Points = -1.0m,
                DefensePointsAllowed35PlusPoints = -4.0m,
                BenchPoints = 0.0m,
                IsActive = true
            };
        }
    }
}

File: the-league/src/types/ScoringSettings.ts
export interface ScoringSettings {
  id: number;
  leagueId: number;
  sport: string;
  
  // Quarterback Scoring
  passingYardsPerPoint: number;
  passingTouchdownPoints: number;
  passingInterceptionPoints: number;
  passingTwoPointConversion: number;
  
  // Rushing Scoring
  rushingYardsPerPoint: number;
  rushingTouchdownPoints: number;
  rushingTwoPointConversion: number;
  
  // Receiving Scoring
  receivingYardsPerPoint: number;
  receivingTouchdownPoints: number;
  receptionPoints: number;
  receivingTwoPointConversion: number;
  
  // General Offensive
  fumbleLostPoints: number;
  
  // Kicker Scoring
  extraPointPoints: number;
  fieldGoal0to39Points: number;
  fieldGoal40to49Points: number;
  fieldGoal50PlusPoints: number;
  missedExtraPointPoints: number;
  missedFieldGoalPoints: number;
  
  // Defense/Special Teams Scoring
  defenseTouchdownPoints: number;
  sackPoints: number;
  interceptionPoints: number;
  fumbleRecoveryPoints: number;
  safetyPoints: number;
  blockedKickPoints: number;
  
  // Defense Points Allowed
  defensePointsAllowed0Points: number;
  defensePointsAllowed1to6Points: number;
  defensePointsAllowed7to13Points: number;
  defensePointsAllowed14to20Points: number;
  defensePointsAllowed21to27Points: number;
  defensePointsAllowed28to34Points: number;
  defensePointsAllowed35PlusPoints: number;
  
  // Additional settings
  benchPoints: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ScoringPreset {
  name: string;
  description: string;
  settings: Partial<ScoringSettings>;
}

export interface ScoringPresetsResponse {
  sport: string;
  presets: ScoringPreset[];
}

export type ScoringCategory = 
  | 'passing'
  | 'rushing' 
  | 'receiving'
  | 'kicking'
  | 'defense'
  | 'general';

export interface ScoringCategoryConfig {
  key: ScoringCategory;
  name: string;
  description: string;
  settings: Array<{
    key: keyof ScoringSettings;
    label: string;
    description: string;
    type: 'number' | 'decimal';
    min?: number;
    max?: number;
    step?: number;
    suffix?: string;
  }>;
}

// Yahoo Fantasy Football Default Values
export const YAHOO_NFL_DEFAULTS: Partial<ScoringSettings> = {
  sport: 'NFL',
  
  // Passing
  passingYardsPerPoint: 25,
  passingTouchdownPoints: 4,
  passingInterceptionPoints: -1,
  passingTwoPointConversion: 2,
  
  // Rushing
  rushingYardsPerPoint: 10,
  rushingTouchdownPoints: 6,
  rushingTwoPointConversion: 2,
  
  // Receiving (Half-PPR)
  receivingYardsPerPoint: 10,
  receivingTouchdownPoints: 6,
  receptionPoints: 0.5,
  receivingTwoPointConversion: 2,
  
  // General
  fumbleLostPoints: -2,
  
  // Kicking
  extraPointPoints: 1,
  fieldGoal0to39Points: 3,
  fieldGoal40to49Points: 4,
  fieldGoal50PlusPoints: 5,
  missedExtraPointPoints: -1,
  missedFieldGoalPoints: -1,
  
  // Defense
  defenseTouchdownPoints: 6,
  sackPoints: 1,
  interceptionPoints: 2,
  fumbleRecoveryPoints: 2,
  safetyPoints: 2,
  blockedKickPoints: 2,
  
  // Defense Points Allowed
  defensePointsAllowed0Points: 10,
  defensePointsAllowed1to6Points: 7,
  defensePointsAllowed7to13Points: 4,
  defensePointsAllowed14to20Points: 1,
  defensePointsAllowed21to27Points: 0,
  defensePointsAllowed28to34Points: -1,
  defensePointsAllowed35PlusPoints: -4,
  
  benchPoints: 0,
  isActive: true
};

// Scoring Categories Configuration
export const SCORING_CATEGORIES: ScoringCategoryConfig[] = [
  {
    key: 'passing',
    name: 'Passing',
    description: 'Quarterback passing statistics',
    settings: [
      {
        key: 'passingYardsPerPoint',
        label: 'Passing Yards',
        description: 'Yards needed for 1 point',
        type: 'number',
        min: 1,
        max: 100,
        suffix: 'yards = 1 pt'
      },
      {
        key: 'passingTouchdownPoints',
        label: 'Passing Touchdown',
        description: 'Points per passing TD',
        type: 'decimal',
        min: 0,
        max: 10,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'passingInterceptionPoints',
        label: 'Interception',
        description: 'Points lost per interception thrown',
        type: 'decimal',
        min: -5,
        max: 0,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'passingTwoPointConversion',
        label: '2-Point Conversion',
        description: 'Points for passing 2-point conversion',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      }
    ]
  },
  {
    key: 'rushing',
    name: 'Rushing',
    description: 'Running back and rushing statistics',
    settings: [
      {
        key: 'rushingYardsPerPoint',
        label: 'Rushing Yards',
        description: 'Yards needed for 1 point',
        type: 'number',
        min: 1,
        max: 50,
        suffix: 'yards = 1 pt'
      },
      {
        key: 'rushingTouchdownPoints',
        label: 'Rushing Touchdown',
        description: 'Points per rushing TD',
        type: 'decimal',
        min: 0,
        max: 10,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'rushingTwoPointConversion',
        label: '2-Point Conversion',
        description: 'Points for rushing 2-point conversion',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      }
    ]
  },
  {
    key: 'receiving',
    name: 'Receiving',
    description: 'Wide receiver and receiving statistics',
    settings: [
      {
        key: 'receivingYardsPerPoint',
        label: 'Receiving Yards',
        description: 'Yards needed for 1 point',
        type: 'number',
        min: 1,
        max: 50,
        suffix: 'yards = 1 pt'
      },
      {
        key: 'receivingTouchdownPoints',
        label: 'Receiving Touchdown',
        description: 'Points per receiving TD',
        type: 'decimal',
        min: 0,
        max: 10,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'receptionPoints',
        label: 'Reception',
        description: 'Points per reception (PPR setting)',
        type: 'decimal',
        min: 0,
        max: 2,
        step: 0.1,
        suffix: 'pts'
      },
      {
        key: 'receivingTwoPointConversion',
        label: '2-Point Conversion',
        description: 'Points for receiving 2-point conversion',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      }
    ]
  },
  {
    key: 'kicking',
    name: 'Kicking',
    description: 'Field goals and extra points',
    settings: [
      {
        key: 'extraPointPoints',
        label: 'Extra Point',
        description: 'Points per successful extra point',
        type: 'decimal',
        min: 0,
        max: 3,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'fieldGoal0to39Points',
        label: 'Field Goal 0-39 yards',
        description: 'Points for field goal 39 yards or less',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'fieldGoal40to49Points',
        label: 'Field Goal 40-49 yards',
        description: 'Points for field goal 40-49 yards',
        type: 'decimal',
        min: 0,
        max: 8,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'fieldGoal50PlusPoints',
        label: 'Field Goal 50+ yards',
        description: 'Points for field goal 50+ yards',
        type: 'decimal',
        min: 0,
        max: 10,
        step: 0.5,
        suffix: 'pts'
      }
    ]
  },
  {
    key: 'defense',
    name: 'Defense/Special Teams',
    description: 'Defensive and special teams statistics',
    settings: [
      {
        key: 'defenseTouchdownPoints',
        label: 'Defensive TD',
        description: 'Points for defensive/ST touchdowns',
        type: 'decimal',
        min: 0,
        max: 10,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'sackPoints',
        label: 'Sack',
        description: 'Points per sack',
        type: 'decimal',
        min: 0,
        max: 3,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'interceptionPoints',
        label: 'Interception',
        description: 'Points per defensive interception',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'fumbleRecoveryPoints',
        label: 'Fumble Recovery',
        description: 'Points per fumble recovery',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'safetyPoints',
        label: 'Safety',
        description: 'Points per safety',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      },
      {
        key: 'blockedKickPoints',
        label: 'Blocked Kick',
        description: 'Points per blocked kick',
        type: 'decimal',
        min: 0,
        max: 5,
        step: 0.5,
        suffix: 'pts'
      }
    ]
  },
  {
    key: 'general',
    name: 'General',
    description: 'Other scoring rules',
    settings: [
      {
        key: 'fumbleLostPoints',
        label: 'Fumble Lost',
        description: 'Points lost per fumble lost',
        type: 'decimal',
        min: -5,
        max: 0,
        step: 0.5,
        suffix: 'pts'
      }
    ]
  }
];